src/libstd/panic.rs

```diff
@@ -0,0 +1,194 @@
+#[unstable(feature = "recover", reason = "recent addition", issue = "27719")]
+#[rustc_on_unimplemented = "the type {Self} contains interior mutability \
+                            and a reference may not be safely transferrable \
+                            across a recover boundary"]
+pub trait NoUnsafeCell {}
+
+/// A simple wrapper around a type to assert that it is panic safe.
+///
+/// When using `recover` it may be the case that some of the closed over
+/// variables are not panic safe. For example if `&mut T` is captured the
+/// compiler will generate a warning indicating that it is not panic safe. It
+/// may not be the case, however, that this is actually a problem due to the
+/// specific usage of `recover` if exception safety is specifically taken into
+/// account. This wrapper struct is useful for a quick and lightweight
+/// annotation that a variable is indeed panic safe.
+///
+/// # Examples
+///
+/// ```
+/// #![feature(recover, std_panic)]
+///
+/// use std::panic::{self, AssertPanicSafe};
+///
+/// let mut variable = 4;
+///
+/// // This code will not compile becuause the closure captures `&mut variable`
+/// // which is not considered panic safe by default.
+///
+/// // panic::recover(|| {
+/// //     variable += 3;
+/// // });
+///
+/// // This, however, will compile due to the `AssertPanicSafe` wrapper
+/// let result = {
+///     let mut wrapper = AssertPanicSafe::new(&mut variable);
+///     panic::recover(move || {
+///         **wrapper += 3;
+///     })
+/// };
+/// // ...
+/// ```
+#[unstable(feature = "recover", reason = "recent addition", issue = "27719")]
+pub struct AssertPanicSafe<T>(T);
+
+// Implementations of the `PanicSafe` trait:
+//
+// * By default everything is panic safe
+// * Mutable references are not panic safe
+// * &T where T contains interior mutability is not panic safe
+// * Types like Mutex/RwLock which are explicilty poisoned are panic safe
+// * Our custom AssertPanicSafe wrapper is indeed panic safe
+impl PanicSafe for .. {}
+impl<'a, T: ?Sized> !PanicSafe for &'a mut T {}
+impl<'a, T: NoUnsafeCell + ?Sized> PanicSafe for &'a T {}
+impl<'a, T: NoUnsafeCell + ?Sized> PanicSafe for *const T {}
+impl<'a, T: NoUnsafeCell + ?Sized> PanicSafe for *mut T {}
+impl<'a, T: PanicSafe> PanicSafe for Unique<T> {}
+impl<T: ?Sized> PanicSafe for Mutex<T> {}
+impl<T: ?Sized> PanicSafe for RwLock<T> {}
+impl<T> PanicSafe for AssertPanicSafe<T> {}
+
+// Pretty simple implementations for the `NoUnsafeCell` marker trait, basically
+// just saying that this is a marker trait and `UnsafeCell` is the only thing
+// which doesn't implement it (which then transitively applies to everything
+// else.
+impl NoUnsafeCell for .. {}
+impl<T: ?Sized> !NoUnsafeCell for UnsafeCell<T> {}
+
+impl<T> AssertPanicSafe<T> {
+    /// Creates a new `AssertPanicSafe` wrapper around the provided type.
+    #[unstable(feature = "recover", reason = "recent addition", issue = "27719")]
+    pub fn new(t: T) -> AssertPanicSafe<T> {
```

Comment:
My feelings on this are:

* `::new` isn't that much longer (5 chars)
* Public fields in tuple structs isn't necessarily idiomatic in libstd
* It's tough for us to predict what kind of subtle interactions we may want in the future (e.g. `PhantomData` for weird variance whatnots). I agree that I'd be *quite* surprised if we ever changed this, but if we do it's probably important.